federation-js/src/joinSpec.ts

```diff
@@ -0,0 +1,96 @@
+import {
+  GraphQLDirective,
+  DirectiveLocation,
+  GraphQLEnumType,
+  GraphQLScalarType,
+  GraphQLString,
+} from 'graphql';
+import { ServiceDefinition } from './composition';
+
+const FieldSetScalar = new GraphQLScalarType({
+  name: 'join__FieldSet',
+});
+
+const EndpointDirective = new GraphQLDirective({
+  name: "join__endpoint",
+  locations: [DirectiveLocation.ENUM_VALUE],
+  args: {
+    serviceName: {
+      type: GraphQLString,
+    },
+    url: {
+      type: GraphQLString,
+    },
+  }
+});
+
+function getJoinGraphEnum(serviceList: ServiceDefinition[]) {
+  return new GraphQLEnumType({
+    name: 'join__Graph',
+    values: Object.fromEntries(
+      serviceList.map((service) => [
+        service.name.toUpperCase(),
```

Comment:
So specifically we might have to do something more special here (or at an earlier place in the process?) to ensure that these values are (a) valid GraphQL names and (b) are unique after upper-casing.

Reply:
(no insight onto whether the proper fix if these fail should be reporting an error vs sanitizing/uniquifying.  we should make sure though that if we uniquify that we do so in a deterministic fashion... though I guess these enum values technically don't need to be stable across composition runs, it would be nice if they were.)

Reply:
Funny, we talked about the uppercasing issue today. Though it didn't cross our mind that the `serviceName` must now be a valid graphql name which is a great consideration. If we're going to go to the length of sanitizing/uniquifying, should we just generate the enum value itself with the knowledge that anything important is captured in the directive on it?

Suppose my services:
`1products`, `2products`, etc.

We _have_ to modify them, is there still a point in preserving some semblance of the original or...? I'm not sure, to be honest!

Reply:
I think having the enum values be something a human can look at in the rest of the schema and have associations with is valuable.

Frankly if we don't want to do sanitizing/uniquifying I'd rather we just go back to providing them as strings to the other directives rather than enums... But I think a light dusting of sanitization and uniquification would be reasonable. Something like:

1: Replace all non-alphanumerics with `_`
2: If it starts with a digit, prefix one `_`
3: (MAYBE, not sure this actually improves things) Upper-case all letters
4: If any string ends with an underscore followed by a string of digits (`_[0-9]+$`), add another underscore at the end.
5: In case of conflicts, postfix all strings with another `_` and a number. Choose the numbers based on the original ordering of the strings before all translations, 1-indexed.

What's the point of step 4? It's so that if you start with `A` `a` `A_1` you end up with `A_1` `A_2` `A_1_` rather than the third one colliding with the first one.

Reply:
Yeah, this is definitely more complicated than I thought it would be. But if we want to use graph names as enum values,   we're going to need at least some variant of this algorithm. Uppercasing isn't strictly needed, but using uppercase for enum values is a strong convention in GraphQL, so it does make the SDL more humanly readable.